qiskit_optimization/algorithms/qrao/magic_rounding.py

```diff
@@ -0,0 +1,448 @@
+                corresponds to the number of shots to use for the corresponding basis in
+                the bases array.
+        """
+        bases_ = [
+            self.rng.choice(2 ** (vars_per_qubit - 1), size=len(q2vars)).tolist()
+            for _ in range(self._shots)
+        ]
+        bases, basis_shots = np.unique(bases_, axis=0, return_counts=True)
+        return bases, basis_shots
+
+    def _sample_bases_weighted(
+        self, q2vars: List[List[int]], expectation_values: List[float], vars_per_qubit: int
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Perform weighted sampling from the expectation values. The goal is to make smarter choices
+        about which bases to measure in using the expectation values.
+
+        Args:
+            q2vars: A list of lists of integers. Each inner list contains the indices of decision
+                variables mapped to a specific qubit.
+            expectation_values: A list of expectation values for each decision variable.
+            vars_per_qubit: The maximum number of decision variables that can be mapped to a
+                single qubit.
+
+        Returns:
+            A tuple containing two arrays:
+                bases: A 2D numpy array of shape (num_bases, num_qubits), where each row
+                corresponds to a basis configuration. Each element of the array is an
+                integer in the range [0, 2 ** (vars_per_qubit - 1) - 1]. The integer
+                represents the index of the basis to measure in for the corresponding
+                qubit.
+                basis_shots: A 1D numpy array of shape (num_bases,), where each element
+                corresponds to the number of shots to use for the corresponding basis in
+                the bases array.
+        """
+        # pylint: disable=C0401
+        # First, we make sure all Pauli expectation values have absolute value
+        # at most 1.  Otherwise, some of the probabilities computed below might
+        # be negative.
+        clipped_expectation_values = np.clip(expectation_values, -1, 1)
+        # basis_probs will have num_qubits number of elements.
+        # Each element will be a list of length 4 specifying the
+        # probability of picking the corresponding magic basis on that qubit.
+        basis_probs = []
+        for dvars in q2vars:
+            if vars_per_qubit == 3:
+                x = 0.5 * (1 - clipped_expectation_values[dvars[0]])
+                y = 0.5 * (1 - clipped_expectation_values[dvars[1]]) if (len(dvars) > 1) else 0
+                z = 0.5 * (1 - clipped_expectation_values[dvars[2]]) if (len(dvars) > 2) else 0
+                # ppp:   mu±   = .5(I ± 1/sqrt(3)( X + Y + Z))
+                # pmm: X mu± X = .5(I ± 1/sqrt(3)( X - Y - Z))
+                # mpm: Y mu± Y = .5(I ± 1/sqrt(3)(-X + Y - Z))
+                # mmp: Z mu± Z = .5(I ± 1/sqrt(3)(-X - Y + Z))
```

Comment:
Is this commented code here going to be deleted? Same below a bit. 

And this seems to be overall another if that is on the range 1 to 3 like I commented on before above, so similar comment applies

Reply:
It's not commented code.  It's equations from the paper that are meant to clarify the lines that follow.

Reply:
Ok, was not sure if it was work in progress or what :) I guess for me not knowing the paper it does not help much - I saw teh code with kinda similar names, quite what this was in relation to that I did not know. I guess perhaps if you have read the paper then you can kinda figure things out and it can be helpful - I can't judge.

Reply:
I can add this comment to clarify what `ppp`, ... are.
```
# In the coefficient of the Pauli operator within the magic bases, 'p' represents a positive sign, while 'm' signifies a negative sign.
# The four combinations of these signs are used to define the quantum system behavior in the context of magic bases.                
```
 I'm wondering whether I add `ppp`, ... to `pylintdict` or not. If we have better names, hmm...

Reply:
I think it might be helpful to have some more info here, as in they are equations from the paper and perhaps the meaning you suggest if that's not so evident from reading the paper. We often end up adding variables we have in comments to the custom dict. Maybe adding hyphens e.g. `p-p-p`, `m-p-m` might avoid that (not sure), if that seems reasonable.

Reply:
Let's add `ppp`,... to `pylintdict`